### PR TITLE
[Disco] Expose `DiscoWorker` and `ndarray_cache_support` in header

### PR DESCRIPTION
**Pull request by LeshengJin**

### CODE REVIEW

File: include/tvm/runtime/disco/builtin.h [hunk 1]
```diff
@@ -51,7 +49,7 @@ NDArray DiscoEmptyNDArray(ShapeTuple shape, DataType dtype, Device device);
  * \param reduce_kind The kind of reduction operation (e.g. sum, avg, min, max)
  * \return The outcome of allreduce
  */
-void AllReduce(NDArray send, ReduceKind reduce_kind, NDArray recv);
```
Comment (junrushao) [inline include/tvm/runtime/disco/builtin.h:54]: Move the definition of `ReduceKind` to this file

File: python/tvm/runtime/disco/process_pool.py [hunk 2]
```diff
@@ -47,9 +47,10 @@ class DiscoPopenWorker:
         The standard error streams handler specified for the popen process.
     """
 
-    def __init__(self, worker_id: int, num_workers: int, stdout=None, stderr=None):
+    def __init__(self, worker_id: int, num_workers: int, entry_point="tvm.exec.disco_worker", stdout=None, stderr=None):
```
Comment (junrushao) [inline python/tvm/runtime/disco/process_pool.py:50]:
```suggestion
    def __init__(self, worker_id: int, num_workers: int, entrypoint="tvm.exec.disco_worker", stdout=None, stderr=None):
```

**Comment by junrushao** [comment]
Instead of moving `worker.h` to `include/tvm/runtime/disco/worker.h`, let's only expose `DiscoWorker` and continue keep `WorkerZeroData` and `DiscoWorkerThread` internal

File: src/runtime/disco/loader.cc [hunk 3]
```diff
@@ -1,380 +1,380 @@
-/*
```
Comment (junrushao) [inline src/runtime/disco/loader.cc:1]: let's keep it as it is

File: src/runtime/relax_vm/ndarray_cache_support.cc [hunk 4]
```diff
@@ -268,6 +267,63 @@ class NDArrayCache {
...
+  /*!
+   * \brief Load parameters from i-th file record and append them.
+   *
+   * \param metadata_json_str The metadata json string.
+   * \param cache_path The cache to path.
+   * \param i The index of file record.
+   * \param device_type The type of device to be loaded.
+   * \param device_id The device id.
+   */
+  static Array<String> LoadFromRecord(const std::string& metadata_json_str,
```
Comment (jinhongyii) [inline src/runtime/relax_vm/ndarray_cache_support.cc:279]: extract into helper method

File: include/tvm/runtime/disco/utils.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef TVM_RUNTIME_DISCO_UTILS_H_
+#define TVM_RUNTIME_DISCO_UTILS_H_
+
+namespace tvm {
+namespace runtime {
+
+/*!
```
Comment (junrushao) [inline include/tvm/runtime/disco/utils.h:25]: move this to builtin.h

File: src/runtime/disco/nccl/nccl.cc [hunk 6]
```diff
@@ -28,7 +28,7 @@
 #include <vector>
 
 #include "../../../support/process_id.h"
-#include "../utils.h"
+#include "../utils.cc"
```
Comment (junrushao) [inline src/runtime/disco/nccl/nccl.cc:31]: do not include a cc file

**Comment by junrushao** [comment]
I just pushed a few refactoring

**Comment by Hzfengsy** [comment]
A minor comment: it would be great to use `NNModule` or `Frontend` as the tag in the title. `SLM` is somehow a name used in MLC-LLM, rather than TVM

**Review by junrushao — APPROVED** [review]

**Pull request merged**